### PR TITLE
Finish flows immediately

### PR DESCRIPTION
**Pull request by nicpottier**
Marks runs as inactive as soon as they reach a terminal node or have a non-matching ruleset.

### CODE REVIEW

File: temba/flows/models.py [hunk 1]
```diff
@@ -555,38 +555,15 @@ def find_and_handle(cls, msg, started_flows=None, voice_response=None):
...
-            # our action set may have disappeared, in that case we just move on, ignoring this flow
-            if action_set:
-                destination = Flow.get_node(flow, action_set.destination, action_set.destination_type)
-                if destination:
-                    flow.add_step(step.run, destination, previous_step=step, arrived_on=timezone.now())
-
-                    # if we pushed them forward, we want to consider any step type
-                    # they just arrived at instead of just rulesets
-                    step_type_filter = None
-
```
Comment (nicpottier) [inline temba/flows/models.py]: This is all fast-forwarding, which we don't do anymore.

File: temba/tests.py [hunk 2]
```diff
@@ -218,8 +227,11 @@ def update_destination_no_check(self, flow, node, destination, rule=None):
             for r in rules:
                 if r.uuid == rule:
                     r.destination = destination
+                    r.destination_type = destination_type
             ruleset.set_rules(rules)
             ruleset.save()
+        else:
+            self.fail("Couldn't find node with uuid: %s" % node)
```
Comment (nicpottier) [inline temba/tests.py]: Some of the tests were passing old UUIDs that didn't match anything.. this makes sure that isn't possible.

File: temba/tests.py [hunk 3]
```diff
@@ -207,9 +207,18 @@ def update_destination(self, flow, source, destination):
...
+        action_destination = Flow.get_node(flow, destination, destination_type)
+        if not action_destination:
+            destination_type = RULE_SET
+            ruleset_destination = Flow.get_node(flow, destination, destination_type)
+            self.assertTrue(ruleset_destination, "Unable to find new destination with uuid: %s" % destination)
+
         actionset = ActionSet.get(flow, node)
         if actionset:
             actionset.destination = destination
+            actionset.destination_type = destination_type
```
Comment (nicpottier) [inline temba/tests.py]: Changing just the destination and not the destination type causes the lookup to fail during flow execution, so we set it here explicitely.

File: temba/flows/tests.py [hunk 4]
```diff
@@ -2806,15 +2809,37 @@ def test_server_runtime_cycle(self):
...
             self.update_destination(flow, group_one_rule.uuid, first_actionset.uuid)
 
+        # add our contact to Group A
+        group_a = ContactGroup.user_groups.create(org=self.org, name="Group A",
+                                                  created_by=self.admin, modified_by=self.admin)
+        group_a.contacts.add(self.contact)
+
         # rule turning back on ourselves
         self.update_destination_no_check(flow, group_ruleset.uuid, group_ruleset.uuid, rule=group_one_rule.uuid)
-        self.send_message(flow, "1", assert_reply=False)
```
Comment (nicpottier) [inline temba/flows/tests.py]: We weren't actually testing anything here.. assert_reply wasn't asserting that there was no reply, and this should actually be throwing.

File: temba/flows/tests.py [hunk 5]
```diff
@@ -2806,15 +2809,37 @@ def test_server_runtime_cycle(self):
...
+        with self.assertRaises(FlowException):
+            self.send_message(flow, "1", assert_reply=False)
+
         flow.delete()
 
         # non-blocking rule to non-blocking rule and back
         flow = self.get_flow('loop_detection')
-        self.update_destination_no_check(flow, name_ruleset.uuid, group_ruleset.uuid, rowan_rule.uuid)
-        self.send_message(flow, "2", assert_reply=False)
```
Comment (nicpottier) [inline temba/flows/tests.py]: Same here.

File: temba/flows/models.py [hunk 6]
```diff
@@ -3030,11 +2991,28 @@ def release(self):
...
+
+            now = timezone.now()
+
+            # mark that we left this step
+            if final_step:
+                final_step.left_on = now
+                final_step.save(update_fields=['left_on'])
+                self.flow.remove_active_for_step(final_step)
+
+            # mark this flow as inactive
```
Comment (norkans7) [inline temba/flows/models.py]: comment update

**Comment by norkans7** [comment]
Looks good to me

**Pull request merged**